File: 0034-find-first-and-last-position-of-element-in-sorted-array/solution.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

int startIdx(int* nums, int n, int target) {
    int left = 0;
    int right = n - 1, val = -1;
    while (left <= right) {
        int mid = (left + right) / 2;
        if (nums[mid] == target) {
            val = mid;
            right = mid - 1;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return val;
}

int endIdx(int* nums, int n, int target) {
    int left = 0;
    int right = n - 1, val = -1;
    while (left <= right) {
        int mid = (left + right) / 2;
        if (nums[mid] == target) {
            val = mid;
            left = mid + 1;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return val;
}
int* searchRange(int* nums, int n, int target, int* returnSize) {
    int* ans = malloc(sizeof(int)*2);
    *returnSize = 2;
    ans[0] = startIdx(nums,n,target);
    ans[1] = endIdx(nums,n , target);
    return ans;
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/solution.java
class Solution {
    private int startValue(int[] nums, int a) {
        int n = nums.length;
        int left = 0, right = n - 1;
        int val = -1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == a) {
                val = mid;
                right = mid - 1;
            } else if (nums[mid] > a) {
                right = mid - 1;
            } else
                left = mid + 1;
        }
        return val;
    }

    private int endValue(int[] nums, int a) {
        int n = nums.length;
        int left = 0, right = n - 1;
        int val = -1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == a) {
                val = mid;
                // right = mid-1;
                left = mid + 1;
            } else if (nums[mid] > a) {
                right = mid - 1;
            } else
                left = mid + 1;
        }
        return val;
    }

    public int[] searchRange(int[] nums, int target) {
        int startIdx = startValue(nums, target);
        int endIdx = endValue(nums, target);
        int[] ans = new int[2];
        ans[0] = startIdx;
        ans[1] = endIdx;
        return ans;
    }
}
